File: server/services/payments.js
/* Payment gateway adapter. PAYMENT_MODE=mock (default) marks payments as paid immediately.
   PAYMENT_MODE=razorpay creates a Razorpay order and verifies the checkout signature. */
const crypto = require('crypto');
const mode = () => (process.env.PAYMENT_MODE || 'mock').toLowerCase();

async function createOrder(amountRupees, receipt) {
  const key = process.env.RAZORPAY_KEY_ID, secret = process.env.RAZORPAY_KEY_SECRET;
  if (!key || !secret) throw new Error('Razorpay keys are not configured');
  const r = await fetch('https://api.razorpay.com/v1/orders', {
    method: 'POST',
    headers: { Authorization: 'Basic ' + Buffer.from(key + ':' + secret).toString('base64'), 'Content-Type': 'application/json' },
    body: JSON.stringify({ amount: Math.round(amountRupees * 100), currency: 'INR', receipt })
  });
  const j = await r.json();
  if (!r.ok) throw new Error((j.error && j.error.description) || 'Payment gateway error');
  return { orderId: j.id, amount: j.amount, keyId: key };
}

function verifySignature(orderId, paymentId, signature, secret = process.env.RAZORPAY_KEY_SECRET) {
  if (!secret || !orderId || !paymentId || !signature) return false;
  const expected = crypto.createHmac('sha256', secret).update(orderId + '|' + paymentId).digest('hex');
  const a = Buffer.from(expected), b = Buffer.from(String(signature));
  return a.length === b.length && crypto.timingSafeEqual(a, b);
}
module.exports = { mode, createOrder, verifySignature };

File: backend-python/app/services/payments.py
"""Payment gateway adapter — port of server/services/payments.js.
PAYMENT_MODE=mock (default) marks payments as paid immediately.
PAYMENT_MODE=razorpay creates a Razorpay order and verifies the checkout signature."""
import base64
import hashlib
import hmac
import os

import httpx

from ..config import get_settings


def mode() -> str:
    return os.environ.get("PAYMENT_MODE", "mock").lower()


async def create_order(amount_rupees: int, receipt: str) -> dict:
    key = os.environ.get("RAZORPAY_KEY_ID")
    secret = os.environ.get("RAZORPAY_KEY_SECRET")
    if not key or not secret:
        raise RuntimeError("Razorpay keys are not configured")
    auth = base64.b64encode(f"{key}:{secret}".encode()).decode()
    async with httpx.AsyncClient() as client:
        r = await client.post(
            "https://api.razorpay.com/v1/orders",
            headers={"Authorization": "Basic " + auth, "Content-Type": "application/json"},
            json={"amount": round(amount_rupees * 100), "currency": "INR", "receipt": receipt},
            timeout=15)
    body = r.json() if r.content else {}
    if r.status_code >= 400:
        raise RuntimeError((body.get("error") or {}).get("description") or "Payment gateway error")
    return {"orderId": body.get("id"), "amount": body.get("amount"), "keyId": key}


def verify_signature(order_id, payment_id, signature, secret: str | None = None) -> bool:
    """HMAC-SHA256 over 'order|payment' — constant-time compare (Node parity)."""
    secret = secret or os.environ.get("RAZORPAY_KEY_SECRET")
    if not secret or not order_id or not payment_id or not signature:
        return False
    expected = hmac.new(secret.encode(), f"{order_id}|{payment_id}".encode(), hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, str(signature))
